pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go

```diff
@@ -913,3 +978,66 @@ func ensureStaticIP(s CloudAddressService, name, serviceName, region, existingIP
 
 	return addr.Address, existed, nil
 }
+
+func (gce *GCECloud) getServiceNetworkTier(svc *v1.Service) (NetworkTier, error) {
+	if !gce.AlphaFeatureGate.Enabled(AlphaFeatureNetworkTiers) {
+		return NetworkTierDefault, nil
+	}
+	tier, err := GetServiceNetworkTier(svc)
+	if err != nil {
+		// Returns an error if the annotation is invalid.
+		return NetworkTier(""), err
+	}
+	return tier, nil
+}
+
+// reconcileFwdRuleNetworkTiers check the network tiers of existing forwarding
+// rule and delete the rule if the tier does not matched the desired tier.
+func reconcileFwdRuleNetworkTiers(s CloudForwardingRuleService, region, name, lbRef string, desiredNetTier NetworkTier) error {
```

Comment:
Using the word `reconcile` seems strange to me here because you're only deleting the forwarding rule/address if the tier needs changed. 

I suggest wripping the two `reconcile` funcs apart and introduce a higher level func which checks tier equality and deletes the resources if necessary.

```go
func deleteWrongNetworkTieredResources(...){ 
  fwdRuleOk, err := fwdRuleNetworkTierEquals(gce, region, name, lbRef, desiredNetTier)
  if err!= nil {
...
} 
if !fwdRuleOk {
  err = s.DeleteRegionForwardingRule(name, region)
}

 addrOk, err := addressNetworkTierEquals(gce, region, name, lbRef, desiredNetTier)
  if err!= nil {
...
} 
if !addrOk {
  err = s.DeleteRegionAddress(name, region)
}

}

Reply:
Agree that the names are confusing. The function was initially intended to do more, but that didn't pan out.
Renamed the functions. I still keep the address and forwarding rule verification in two functions to make testing easier.